### DISCUSSION TITLE
Unexpected Results: Higher cls Weight Leads to Worse Classification – Why?

### DISCUSSION THREAD

**Discussion by Lingpheng** [ultralytics/ultralytics#20899] [Q&A]
Hi, I am currently training a model to detect multiple objects. The exact positions of the bounding boxes are not very important to me — I care more about the accurate classification of different objects. So, I tried increasing cls loss weight, while keeping the weights for box and dfl  unchanged.

I compared two values for the cls weight: on the left, cls=0.8, and on the right, cls=1.5. All other training settings were the same.

Here are the training and validation results:

![Cls](https://github.com/user-attachments/assets/b7cfc889-7b9a-4c95-95aa-7878c14e3655)

And here are the corresponding test results:

![Resultat of Test](https://github.com/user-attachments/assets/e8e3a9d6-7d4c-4a10-819a-31475c7dc03b)

From these results, it seems that cls=0.8 performs better — the final classification loss is 0.75 when cls=0.8, compared to 1.5 when cls=1.5.

So I have two questions:

    1. Isn’t it supposed to be that a higher cls weight makes the model focus more on improving classification performance? Why does cls=0.8 
        perform better than cls=1.5 in my case?

    2. How should I best set the weights for cls, box, and dfl if my main goal is to train a model that emphasizes classification accuracy more 
        than localization?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Lingpheng, thank you for sharing your detailed observations and results with Ultralytics 🚀! For users interested in optimizing loss weights, we recommend reviewing the [Docs](https://docs.ultralytics.com/), including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. These resources cover many of the most common questions about custom training and model configuration.

If this is a 🐛 Bug Report, please include a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate further.

For custom training ❓ questions like yours, providing as much context as possible—including dataset samples, relevant config files, and complete training logs—will help our team assist you effectively. Also, please review our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to ensure you're following best practices.

Join the Ultralytics community where it suits you best! For real-time chat, visit [Discord](https://discord.com/invite/ultralytics) 🎧. For thoughtful discussions, check out [Discourse](https://community.ultralytics.com/). Or connect with others on our [Subreddit](https://reddit.com/r/Ultralytics).

## Upgrade

To make sure you are working with the latest features and fixes, upgrade to the newest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these verified environments (with dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response 🤖. An Ultralytics engineer will review your discussion and assist you soon!

**Reply by glenn-jocher** [reply]
Higher classification weights don't always improve classification performance because loss components work together - over-weighting `cls` can actually hurt localization quality, which in turn degrades classification accuracy since the model needs good bounding box predictions to classify objects effectively.

The issue is that when `cls=1.5` forces the model to over-prioritize classification loss, it may sacrifice localization quality, leading to worse overall performance. A balanced approach typically works better.

For classification-focused training, try gradually increasing `cls` weight (start with 1.0-1.2) while slightly reducing `box` weight, and monitor validation metrics. You might also consider using a smaller model size to naturally emphasize classification over precise localization, or experiment with confidence thresholds during inference to filter predictions.

**Comment by Lingpheng** [comment]
Thanks! I see—it's important to balance the three parameters. I'll try using the n model and start experimenting with cls=1.0. Appreciate the help!

**Reply by glenn-jocher** [reply]
You're welcome! That sounds like a solid plan - the n model is perfect for testing and cls=1.0 is a great starting point. Good luck with your experiments, and feel free to share your results!

**Comment by Y-T-G** [comment]
The loss would be higher because the weights are multiplied with the actual loss value. That's what you see in the log. Higher weights equals higher value of loss. They are not comparable without normalizing first with respect to the weights

**Reply by Lingpheng** [reply]
Thanks for your answer — you're right!

I looked into the code and found the v8DetectionLoss class in loss.py, where I saw the following lines:

```
loss[0] *= self.hyp.box  # box gain  
loss[1] *= self.hyp.cls  # cls gain  
loss[2] *= self.hyp.dfl  # dfl gain
```

So the final loss values are actually the raw computed losses multiplied by the weights we set for box, cls, and dfl during training. A larger cls weight leads to a higher classification loss value, which in turn results in a larger gradient for classification during backpropagation.

That means it's not meaningful to directly compare the loss values in the training logs of different models, since those values have already been scaled by the respective hyperparameters.

This makes things much clearer now!

**Reply by glenn-jocher** [reply]
Exactly right! You've identified the key insight - the logged loss values are the weighted losses, not the raw unscaled losses. This is why comparing loss values across different weight configurations can be misleading.

For meaningful comparison, you'd want to look at the actual performance metrics like mAP, precision, and recall on your validation set, which reflect the model's true classification and detection abilities rather than the scaled loss components used for gradient computation.

**Reply by Lingpheng** [reply]
So would `box=7.5, cls=0.5, dfl=1.5 and box=15, cls=1, dfl=3` behave the same, since the relative weights are the same? As long as the ratios between the losses stay constant, the optimization direction shouldn't change, right?

**Reply by glenn-jocher** [reply]
Theoretically yes, those two configurations should behave very similarly since the ratios are identical (15:1:3). The optimization direction should be essentially the same because the relative importance between box, cls, and dfl losses remains constant.

However, there might be subtle differences due to the absolute gradient magnitudes being 2x larger in the second case, which could interact slightly differently with your learning rate and optimizer dynamics. In practice, the first configuration (smaller absolute values) is generally preferred to avoid potential numerical issues.
